Enable creation and transfer of ScriptModule over RPC

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48339 RRef proxy support for ScriptModule methods
* **#48293 Enable creation and transfer of ScriptModule over RPC**

This PR enables a `ScriptModule` to be created on a remote worker, retrieve it to current device, and run methods on it remotely.

In order to do this, we define custom pickling for `ScriptModule` in our `InternalRPCPickler`. The pickling basically uses torch.save and torch.load to save/recover the ScriptModule.

We test that we can create and retrieve a ScriptModule with rpc_sync, and create RRef to ScriptModule with rpc.remote. We can also run remote methods on the rref and transfer it to current worker.

Although we can run methods remotely on the RRef to ScriptModule, this does not currently work with RRef helper, support is added in the next (wip) PR.

Future PRs will address:
1. Calling local ScriptModule method remotely 
2. Calling `to_here()` on ScriptModule in JIT (currently disabled, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_impl.cpp#L147). 

Differential Revision: [D25107773](https://our.internmc.facebook.com/intern/diff/D25107773/)

rohan-varma committed:
Enable creation and transfer of ScriptModule over RPC

This PR enables a `ScriptModule` to be created on a remote worker, retrieve it to current device, and run methods on it remotely.

In order to do this, we define custom pickling for `ScriptModule` in our `InternalRPCPickler`. The pickling basically uses torch.save and torch.load to save/recover the ScriptModule.

We test that we can create and retrieve a ScriptModule with rpc_sync, and create RRef to ScriptModule with rpc.remote. We can also run remote methods on the rref and transfer it to current worker.

Although we can run methods remotely on the RRef to ScriptModule, this does not currently work with RRef helper, filed to track that.

Differential Revision: [D25107773](https://our.internmc.facebook.com/intern/diff/D25107773/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable creation and transfer of ScriptModule over RPC"


This PR enables a `ScriptModule` to be created on a remote worker, retrieve it to current device, and run methods on it remotely.

In order to do this, we define custom pickling for `ScriptModule` in our `InternalRPCPickler`. The pickling basically uses torch.save and torch.load to save/recover the ScriptModule.

We test that we can create and retrieve a ScriptModule with rpc_sync, and create RRef to ScriptModule with rpc.remote. We can also run remote methods on the rref and transfer it to current worker.

Although we can run methods remotely on the RRef to ScriptModule, this does not currently work with RRef helper, filed to track that.

Future PRs will address:
1. Calling local ScriptModule method remotely 
2. Calling `to_here()` on ScriptModule in JIT (currently disabled, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_impl.cpp#L147). 

Differential Revision: [D25107773](https://our.internmc.facebook.com/intern/diff/D25107773/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Enable creation and transfer of ScriptModule over RPC"


This PR enables a `ScriptModule` to be created on a remote worker, retrieve it to current device, and run methods on it remotely.

In order to do this, we define custom pickling for `ScriptModule` in our `InternalRPCPickler`. The pickling basically uses torch.save and torch.load to save/recover the ScriptModule.

We test that we can create and retrieve a ScriptModule with rpc_sync, and create RRef to ScriptModule with rpc.remote. We can also run remote methods on the rref and transfer it to current worker.

Although we can run methods remotely on the RRef to ScriptModule, this does not currently work with RRef helper, filed to track that.

Future PRs will address:
1. Calling local ScriptModule method remotely 
2. Calling `to_here()` on ScriptModule in JIT (currently disabled, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_impl.cpp#L147). 

Differential Revision: [D25107773](https://our.internmc.facebook.com/intern/diff/D25107773/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks great, thanks! Looks like some CI and linting are failing, can you fix them before landing?

rohan-varma committed:
Update on "Enable creation and transfer of ScriptModule over RPC"


This PR enables a `ScriptModule` to be created on a remote worker, retrieve it to current device, and run methods on it remotely.

In order to do this, we define custom pickling for `ScriptModule` in our `InternalRPCPickler`. The pickling basically uses torch.save and torch.load to save/recover the ScriptModule.

We test that we can create and retrieve a ScriptModule with rpc_sync, and create RRef to ScriptModule with rpc.remote. We can also run remote methods on the rref and transfer it to current worker.

Although we can run methods remotely on the RRef to ScriptModule, this does not currently work with RRef helper, filed to track that.

Future PRs will address:
1. Calling local ScriptModule method remotely 
2. Calling `to_here()` on ScriptModule in JIT (currently disabled, see https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/rref_impl.cpp#L147). 

Differential Revision: [D25107773](https://our.internmc.facebook.com/intern/diff/D25107773/)

[ghstack-poisoned]
